### DISCUSSION TITLE
Update documents by filter

### DISCUSSION THREAD

**Discussion by MrAtiebatie** [meilisearch/product#695] [Feedback & Feature Proposal]
It would be great to be able to update multiple documents based on a filter. Especially when you're using distinctAttribute it would be nice to update all documents with product_id = 2. Now I have to save all the document primary keys separately and update them individually.

**Comment by irevoire** [comment]
Hello @MrAtiebatie,

What kind of operation would you like to do on your documents?
Could you share some kind of API?

**Reply by MrAtiebatie** [reply]
For example the jackets displayed in the meilisearch documentation: https://www.meilisearch.com/docs/learn/configuration/distinct#example 

It would be nice if I could bulk update the price of all 3 jackets at once. The syntax could be:

```
$client->index('products')->update([
    'price' => 100,
], [
    'filter' => ['product_id = 2'],
]);
```

**Reply by flevi29** [reply]
Basically I'd like to set `verified` of every document with a specific `library_id` to `false`.

I've already detailed this on Discord, but here is the current implementation:

```typescript
async function setLibraryDocumentsToUnverified(
    library_id: number,
    signal?: AbortSignal,
  ): Promise<number> {
    const enqueuedTasks: EnqueuedTask[] = [];
    let page = 0,
      results: Pick<SEDocument, "objectID">[] | null = null;

    for (;;) {
      signal?.throwIfAborted();
      results = (
        await this.#index.getDocuments({
          filter: `${"library_id" satisfies KeyofSEDocument} = ${library_id}`,
          fields: ["objectID"] satisfies KeyofSEDocument[],
          offset: page * 200,
          limit: (page += 1) * 200,
        })
      ).results;

      if (results.length === 0) {
        break;
      }

      enqueuedTasks.push(
        await this.#index.updateDocuments(
          results.map(
            (v): Pick<SEDocument, "objectID" | "verified"> => ({
              ...v,
              verified: false,
            }),
          ),
        ),
      );
    }

    const tasks = await this.waitForEnqueuedTasks(enqueuedTasks);
    return tasks
      .map((task) => task.details.indexedDocuments!)
      .reduce((previousValue, currentValue) => {
        previousValue += currentValue;
        return previousValue;
      }, 0);
  }
```

This could be so much simpler. All of this could also be one task, one that locks all other modifications, so it would be safer and more efficient.

The reason I'm doing this, is so that later I can remove all unverified documents with a query. It is required for syncing an external database.

**Reply by flevi29** [reply]
I'm also noticing that a simple boolean value switch of 74000 documents is taking quite a lot of time and resources. I'm sure the situation described above doesn't help, but this is possibly another topic.

**Reply by irevoire** [reply]
Huum, ok, interesting.

So, you only need to update all the values to a single value.
You never need to update the values based on a function or something?
Like, don’t you want to increment the popularity of your documents by one, or update the price of something based on an HTTP call?

I’m just trying to think of how general our solution should be 🤔 

> I'm also noticing that a simple boolean value switch of 74000 documents is taking quite a lot of time and resources.

We worked on this a lot this sprint and hope it’ll be fixed for the next release 🤞 
It should at least improve a lot.

**Reply by flevi29** [reply]
I, for one, only need to update all documents matching a filter, and update them all with the same value. Much like how [typesense update by query](https://typesense.org/docs/0.25.1/api/documents.html#update-by-query) works, if I understand it correctly. I am sure someone out there might want more advanced features, but this seems the simplest case scenario.

**Reply by MrAtiebatie** [reply]
I have more than 200k documents so it's not a very nice solution. I don't understand how this feature is implemented for deletion but not for updating. Would love to have this :-)

**Reply by flevi29** [reply]
It's a lot simpler I assume when it comes to deletion. Not being able to do this by query is one thing I am willing to let go, but there's a far bigger performance issue, that Meilisearch is aware of and they're working on it: (go and upvote it) https://github.com/meilisearch/meilisearch/issues/4071

Currently my server is stalling, and Meilisearch is timing out, with 30 minutes of time out. None of my procedures complete successfully because of this. What was meant to be a very efficient way of keeping track of what to delete became a major bottle neck once I switched to Meilisearch.
